src/Juvix/Core/Erasure/Algorithm.hs

```diff
@@ -40,28 +54,35 @@ eraseTerm parameterisation term usage ty =
       Core.PrimTy _ → throw @"erasureError" Unsupported
       Core.Pi _ _ _ → throw @"erasureError" Unsupported
       Core.Lam name body → do
+        -- The type must be a dependent function.
         let Core.Pi argUsage varTy retTy = ty
-            bodyUsage = Core.SNat 1
+        -- TODO: Is this correct?
+        let bodyUsage = Core.SNat 1
         ty ← eraseType parameterisation varTy
         modify @"typeAssignment" (Map.insert name ty)
-        -- TODO resTy is a function, which we must deal with.
+        let varTyIR = IR.cEval parameterisation (hrToIR varTy) []
+        modify @"context" ((:) (IR.Global (show name), (argUsage, varTyIR)))
         body ← eraseTerm parameterisation body bodyUsage retTy
         -- If argument is not used, just return the erased body.
-        if usage <> argUsage == Core.SNat 0
-          then pure body
-          else-- Otherwise, if argument is used, return a lambda function.
-
-            pure (Erased.Lam name body)
+        -- Otherwise, if argument is used, return a lambda function.
+        pure (if usage <.> argUsage == Core.SNat 0 then body else Erased.Lam name body)
       Core.Elim elim →
         case elim of
           Core.Var n → pure (Erased.Var n)
           Core.Prim p → pure (Erased.Prim p)
           Core.App f x → do
-            -- TODO Find type of f, if f uses x erase to app else erase to f.
-            f ← eraseTerm parameterisation (Core.Elim f) undefined undefined
-            -- TODO Find type of x (determined by f arg type & usage).
-            x ← eraseTerm parameterisation x undefined undefined
-            pure (Erased.App f x)
+            let IR.Elim fIR = hrToIR (Core.Elim f)
+            context ← get @"context"
+            case IR.iType0 parameterisation context fIR of
+              Left err → throw @"erasureError" (InternalError (show err <> " while attempting to erase " <> show f))
+              Right (fUsage, fTy) → do
+                let fty@(Core.Pi argUsage fArgTy _) = irToHR (IR.quote0 fTy)
+                f ← eraseTerm parameterisation (Core.Elim f) fUsage fty
+                if argUsage == Core.SNat 0
```

Comment:
changes seem fine, however I'm curious about this if statement

Reply:
See the language reference, page 30:

![2019-11-18-133420_459x77_scrot](https://user-images.githubusercontent.com/24760279/69052776-254c6080-0a08-11ea-964b-d2171f736fab.png)